modules/social_features/social_user/src/Plugin/Action/SocialRemoveRoleUser.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+namespace Drupal\social_user\Plugin\Action;
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\social_user\UserRoleHelper;
+use Drupal\user\Plugin\Action\ChangeUserRoleBase;
+use Drupal\user\RoleInterface;
+
+/**
+ * Removes a role from a user but restricted by role.
+ *
+ * @Action(
+ *   id = "social_user_remove_role_action",
+ *   label = @Translation("Remove a role from the selected users"),
+ *   type = "social_user"
+ * )
+ */
+class SocialRemoveRoleUser extends ChangeUserRoleBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute($account = NULL) {
+    $rid = $this->configuration['rid'];
+    // Skip removing the role from the user if they already don't have it.
+    if ($account !== FALSE && $account->hasRole($rid)) {
+      // For efficiency manually save the original account before applying
+      // any changes.
+      $account->original = clone $account;
+      $account->removeRole($rid);
+      $account->save();
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function buildConfigurationForm(array $form, FormStateInterface $form_state) {
+    /** @var \Drupal\social_user\UserRoleHelper $roles */
+    $roles = UserRoleHelper::alterAddOrRemoveRoleOptions(user_role_names(TRUE), 'sitemanager', ['administrator']);
```

Comment:
@frankgraave This looks like we're hardcoding something for sitemanagers but I'm having a hard time figuring out what. Could you explain? : ) What is the helper function doing? Why do we pass it site manager and not content manager? Does this still work of the "Assign X" permissions are different or if there are different roles on the platform?

It may also be clearer if there's some code duplication in the `SocialRemoveRoleUser`/`SocialAddRoleUser` classes instead of a helper that's only used twice.

Reply:
Hi @Kingdutch SM's were able to grant the administrator role to themself. Therefore, as described in the doc comment of the helper function, the 'sitemanager' role is used as an (flexible) check to determine if certain options should or should not be removed from an options list. Open Social always comes with a sitemanager - as a trusted role - if I'm correct. However, there was a scenario where multiple sitemanagers were granting themself the administrator role, which caused a lot of trouble in the platform.

At first, there are indeed only two places where this needs to be applied, but I wanted to keep it DRY. The idea is that if there were other places that need to filter out certain options by role, this helper could be used.

Reply:
It just feels like this helper function is a bit too specifically unspecific.
I can imagine you want to re-use code to say, ok you know the administrator role is something nobody should ever be able to give themselves or to others if you are not one yourself.  
At least my understanding that is what is what we are trying to achieve.

What I would opt for is to not perse do this with a helper function in the current format, i doubt anyone would reuse this and it might be not flexible enough or not clear enough how to make it flexible. 
Alternatively you can create permissions that are saying "allow to assign or revoke the X role". 
So administrator can assign or revoke this always, 
Site manager can assign content manager and SM roles
and you can ship with this by default and check for those permissions?

Reply:
Ah good argument against introducing the abstraction here is AHA Programming: "Avoid Hasty Abstractions".

This is an interesting read by the person who coined the term: https://kentcdodds.com/blog/aha-programming